File: Unit2-07/Typescript/Vehicle.ts
/**
* This class creates a vehicle
*
* By:      Evan Beaudoin
* Version: 1.0
* Since:   2024-05-01
*/

export class Vehicle {
    // Class variables
  
    // Vehicle colour
    protected colour: string
    // Current speed of the vehicle
    protected speed: number
    // Maximum speed of the vehicle
    protected maxSpeed: number
  
    // Creates the base vehicle
    public constructor(initialColour: string, initialMaxSpeed: number) {
      this.colour = initialColour
      this.speed = 0
      this.maxSpeed = initialMaxSpeed
    }
  
    // Gets the colour of the vehicle
    public get colour(): string {
      return this.colour
    }
  
    // Gets the current speed of the vehicle
    public get speed(): number {
      return this.speed
    }
  
    // Gets the maximum speed of the vehicle
    public get maxSpeed(): number {
      return this.maxSpeed
    }
  
    // Creates a string about the status of the car
    public get status(): string {
      let statusString = 'Status:\n'
      statusString = ` -> Speed: ${this.speed}\n`
      statusString += ` -> MaxSpeed: ${this.maxSpeed}\n`
      statusString += ` -> Color: ${this.colour}`
  
      return statusString
    }
  
    // Sets the colour of the vehicle
    public set colour(newColour: string) {
      this.colour = newColour
    }
  
    // Accelerates the car, changing the speed
    public accelerate(accelerationPower: number, accelerationTime: number) {
      if (accelerationPower > 0 && accelerationTime > 0) {
        this.speed = (accelerationPower * accelerationTime) + this.speed
        if (this.speed > this.maxSpeed) {
          this.speed = this.maxSpeed
        }
      } else {
        console.log("Acceleration power and time must be positive numbers.")
      }
    }
  
    // Brakes the car, changing the speed
    public brake(brakePower: number, brakeTime: number) {
      if (brakePower > 0 && brakeTime > 0) {
        this.speed = this.speed - (brakePower * brakeTime)
        if (this.speed < 0) {
          this.speed = 0
        }
      } else {
        console.log("Brake power and time must be positive numbers.")
      }
    }
  }

File: Unit2-07/Java/Vehicle.java
/*
* This class creates a vehicle
*
* @author  Evan Beaudoin
* @version 1.0
* @since   2024-04-12
*/

/**
 * This is the class.
*/
public class Vehicle {
    /**
     * Variable for the class.
     *
     * @param colour of the vehicle
    */
    private String colour = "";

    /**
     * Variables for the class.
     *
     * @param speed of the vehicle
    */
    private double speed;

    /**
     * Variables for the class.
     *
     * @param maxSpeed of the vehicle
    */
    private double maxSpeed;

    /**
     * Creates the base vehicle.
     *
     * @param initialColour the base colour
     * @param initialMaxSpeed the maximum speed of the vehicle
    */
    public Vehicle(String initialColour, double initialMaxSpeed) {
        this.colour = initialColour;
        this.speed = 0;
        this.maxSpeed = initialMaxSpeed;
    }

    /**
     * Gets the colour of the vehicle.
     *
     * @return the current colour
    */
    public String getColour() {
        return this.colour;
    }

    /**
     * Gets the speed of the vehicle.
     *
     * @return the current speed
    */
    public double getSpeed() {
        return this.speed;
    }

    /**
     * Gets the maximum speed of the vehicle.
     *
     * @return the maximum speed
    */
    public double getMaxSpeed() {
        return this.maxSpeed;
    }

    /**
     * Creates a string about the status of the vehicle.
     *
     * @return the string about the vehicle's status
    */
    public String getStatus() {
        final String breakLine = "\n";
        String status = " -> Speed: " + this.speed + breakLine;
        status += " -> MaxSpeed: " + this.maxSpeed + breakLine;
        status += " -> Color: " + this.colour + breakLine;

        return status;
    }

    /**
     * Sets the colour of the vehicle.
     *
     * @param newColour of the vehicle
    */
    public void setColour(String newColour) {
        this.colour = newColour;
    }

    /**
     * Sets the speed of the vehicle.
     *
     * @param newSpeed of the vehicle
    */
    protected void setSpeed(double newSpeed) {
        this.speed = newSpeed;
    }

    /**
     * Accelerates the vehicle, which changes the speed.
     *
     * @param accelerationPower power of the acceleration
     * @param accelerationTime time of the acceleration
    */
    public void accelerate(double accelerationPower, double accelerationTime) {
        if (accelerationPower > 0 && accelerationTime > 0) {
            this.speed = (accelerationPower * accelerationTime) + this.speed;
            if (this.speed > this.maxSpeed) {
                this.speed = this.maxSpeed;
            }
        } else {
            System.out.print("Acceleration power and "
                                + "time must be positive numbers.");
        }
    }

    /**
     * Brakes the vehicle, which changes speed.
     *
     * @param brakePower power of the brake
     * @param brakeTime time of the brake
    */
    public void brake(double brakePower, double brakeTime) {
        if (brakePower > 0 && brakeTime > 0) {
            this.speed = this.speed - (brakePower * brakeTime);
            if (this.speed < 0) {
                this.speed = 0;
            }
        } else {
            System.out.print("Brake power and time must be positive numbers.");
        }
    }
}
